core/src/main/scala-2.13+/cats/data/NonEmptyLazyList.scala

```diff
@@ -458,16 +458,14 @@ sealed abstract private[data] class NonEmptyLazyListInstances extends NonEmptyLa
 
       def extract[A](fa: NonEmptyLazyList[A]): A = fa.head
 
-      def nonEmptyTraverse[G[_]: Apply, A, B](fa: NonEmptyLazyList[A])(f: A => G[B]): G[NonEmptyLazyList[B]] =
-        Foldable[LazyList]
-          .reduceRightToOption[A, G[LazyList[B]]](fa.tail)(a => Apply[G].map(f(a))(LazyList.apply(_))) { (a, lglb) =>
-            Apply[G].map2Eval(f(a), lglb)(_ +: _)
+      def nonEmptyTraverse[G[_]: Apply, A, B](fa: NonEmptyLazyList[A])(f: A => G[B]): G[NonEmptyLazyList[B]] = {
+        def loop(head: A, tail: LazyList[A]): Eval[G[NonEmptyLazyList[B]]] =
+          tail.headOption.fold(Eval.now(Apply[G].map(f(head))(NonEmptyLazyList(_)))) { h =>
+            Apply[G].map2Eval(f(head), Eval.defer(loop(h, tail.tail)))((b, acc) => b +: acc)
           }
```

Comment:
There is a bit of LISP here. This code may be more readable as follows: 

```suggestion
          val fh = f(head)
          tail match {
            case _ if tail.isEmpty => 
              Eval.now(Apply[G].map(fh)(NonEmptyLazyList(_)))
            case h #:: ttail => 
              Apply[G].map2Eval(fh, Eval.defer(loop(h, ttail)))(_ +: _)
          }
```
The changes are: 

- Extract `f(head)`, which is eagerly evaluated in both branches, as a `val`.
- Replace the `Option.fold` by a pattern-match.
- Replace the cases of `headOption` by two cases on the `tail` lazy list itself. For the first case, there seems not to be any case-object, like `Nil`, for the empty list. [Discussion](https://users.scala-lang.org/t/pattern-matching-lazylist/5105).
- In the second case, abbreviate `(b, acc) => b +: acc` by underscores `_ +: _`.

Reply:
Hello, I see what you mean there. 

On my part, usually I try to avoid pattern match (it's slow for the newer object allocations)